### ISSUE TITLE
Submission: zstd

### ISSUE THREAD

**Issue by felixhandte** [r-lyeh/single_file_libs#213]
[Zstd](https://github.com/facebook/zstd) is a widely-deployed, state-of-the-art compression algorithm. Although the library is developed in many source files (so as to, among other things, allow inclusion of different subsets of functionality), there is [tooling](https://github.com/facebook/zstd/tree/dev/contrib/single_file_libs) to create a single compilation unit version of the library. Admittedly, it's not tiny, but it depends only on `libc`, and even that dependency is [optional](https://github.com/facebook/zstd/tree/dev/contrib/freestanding_lib).

**Comment by r-lyeh** [comment]
I think Zstd no longer supports that single compilation unit mode.

**Comment by forksnd** [comment]
> I think Zstd no longer supports that single compilation unit mode.

It should. The links changed, new instructions are here: https://github.com/facebook/zstd/tree/dev/build/single_file_libs

**Comment by r-lyeh** [comment]
Ouch I see. so on Windows I need either to install python3 (which is not a big deal, but hey...) or use some other bash env that I ofc need to install or just run it in a macos/linux shell.

I am not sure why this is not shipped as a Github action or artifact that does the amalgamation automatically, given the fact zstd is so popular and the project is backed up by facebook even. They should have the funds. [/rant off]

I'll add the repo nontheless thanks. It's probably only me working on Windows :D

**Comment by nothings** [comment]
You can run this project however you want, but just to note, I had a general policy of not accepting amalgamated builds, with the premise being that you want the source to be in the same form that the developer maintains, not in some derived form. (Although, really, to some extent, this might have just been an excuse to forbid overly-large libraries.)

And even if I had accepted amalgamated libraries I definitely wouldn't have accepted libraries that weren't shipped amalgamated, but required you to make them yourself, since one of the things I personally think is important to single-file-libraries is ease of integration.

But like I said, do what you prefer. But be advised that the FAQ mentions this specifically for sqlite, though, so you may want to change that if you change the policy.

**Comment by r-lyeh** [comment]
My PD-biased brain totally agrees on your key points. In an ideal world to me, all libraries and contributions in this list should be PD + 1-file header + written in pure C. Then I remind myself that we cannot curate & verify every single contribution and I'd like also the list to be a community effort where everyone could submit their own personal and/or preferred libs if they consider it could benefit the gamedev scene. I honestly don't know where to draw the line cos strictly speaking there are already many exceptions to the rules: some libraries are C++ only, a lot of them are MIT/BSD or GPL licensed, some are +4 files large, others are not cross-platform, or not amalgamated at all, etc. Each library has some kind of manual verification process that can be very exhausting and time-consuming in the long-run, you know that better than anyone of us in here. So, I think I could just leave the rules as a soft requirement and try to merge as many contributions as possible, as long as those meet *some-but-not-all* the cited requirements. Some day maybe, we'll be able to curate them one by one and/or be more selective when the list becomes a messy place. But for now that's only my thoughts about the short-term plans, and I'm totally open to suggestions and comments ofc :D

**Comment by r-lyeh** [comment]
PS: I think I'll make a separate, golden, list on top above all the rest where the pure C, single-file, PD libs that meet (y)our original criteria can rest and be highlighted and bold visible

**Comment by nothings** [comment]
> Each library has some kind of manual verification process that can be very exhausting and time-consuming in the long-run, you know that better than anyone of us in here.

Yep, I almost mentioned exactly this in my earlier comment. I absolutely mean it when I say you should choose for yourself how you want to go about it.

I worry that a list without reasonably enforced constraints isn't very useful to anyone, but then again, a list that never gets updated isn't very useful to anyone either.

**Comment by forksnd** [comment]
> I am not sure why this is not shipped as a Github action or artifact that does the amalgamation automatically, given the fact zstd is so popular and the project is backed up by facebook even. They should have the funds. [/rant off]

They have given their reasoning here: https://github.com/facebook/zstd/issues/305#issuecomment-830190188
Not sure if I agree with it, but I suppose it is what it is.

**Comment by r-lyeh** [comment]
I've already separated the list into two tiers. First tier matches Sean's vision more accurately and second tier meets the expectations of most devs which are fine with soeme other licensing terms and/or using pure C++ libs and/or multi-files. However, I'll be doing a tertiary tier for those libraries that are blatantly different from what you'd expect to find in this listing: zstd (because you have to build the amalgmation yourself in a bash env), dotheworld (because you can only find the amalgamation in the release packages), cpuid (because the license can nowhere be found unless you navigate to a specific github issue), or sqlite or my own engine (because their amalgamation is computer generated and you cannot really inspect the header at all; it is way too large).

Hope that sorts everything out

**Comment by nothings** [comment]
It *might* be better if it were all one list but the exceptions were tagged in some way, because it's kind of a UX nightmare to have to check multiple lists (most people probably won't realize they should, and therefore won't), but maybe it's difficult to come up with a "UI" for displaying that cleanly.

**Comment by r-lyeh** [comment]
Merged ZSTD into a tier 3 list
